source/common/network/connection_impl.cc

```diff
@@ -254,13 +265,51 @@ void ConnectionImpl::write(Buffer::Instance& data) {
     // ever changed, read the comment in Ssl::ConnectionImpl::doWriteToSocket() VERY carefully.
     // That code assumes that we never change existing write_buffer_ chain elements between calls
     // to SSL_write(). That code will have to change if we ever copy here.
-    write_buffer_.move(data);
+    write_buffer_->move(data);
+    checkForHighWatermark();
+
     if (!(state_ & InternalState::Connecting)) {
       file_event_->activate(Event::FileReadyType::Write);
     }
   }
 }
 
+void ConnectionImpl::setWriteBufferWatermarks(size_t new_low_watermark, size_t new_high_watermark) {
+  ENVOY_CONN_LOG(trace, "Setting watermarks: {} {}", *this, new_low_watermark, new_high_watermark);
+  ASSERT(new_low_watermark < new_high_watermark);
+
+  high_watermark_ = new_high_watermark;
+  low_watermark_ = new_low_watermark;
+
+  checkForLowWatermark();
+  checkForHighWatermark();
+}
+
+void ConnectionImpl::checkForLowWatermark() {
+  if (!above_high_watermark_called_ || write_buffer_->length() >= low_watermark_) {
+    return;
+  }
+  ENVOY_CONN_LOG(trace, "onBelowWriteBufferLowWatermark", *this);
+
+  above_high_watermark_called_ = false;
+  for (ConnectionCallbacks* callback : callbacks_) {
+    callback->onBelowWriteBufferLowWatermark();
+  }
+}
+
+void ConnectionImpl::checkForHighWatermark() {
+  if (above_high_watermark_called_ || high_watermark_ == 0 ||
```

Comment:
There's some asymmetry here that is probably fine, but might be worth calling out somewhere. You'll never receive an `onBelowWBLowWatermark()` unless `onAboveWBHighWatermark()` has been called, but the other direction doesn't hold.